### PR TITLE
Sheeping the Subobjects

### PR DESCRIPTION
**Pull request by lomky**
Making the Patient subobjects polymorphic, feeds into #1362

### CODE REVIEW

**Comment by lomky** [comment]
Hacked together a working polymorphic `call` at the model level. Thank you @colinxfleming for the unsticking. 

We decided that there is no real need for the controller to care about the polymorphic nature of these objects, as archive patient objects are accessed via export only.

**Comment by lomky** [comment]
This is ready for review! Could I get eyes from @colinxfleming or @DarthHater, or whoever else feels comfy in rails guts?

File: test/controllers/external_pledges_controller_test.rb [hunk 1]
```diff
@@ -37,7 +37,12 @@ class ExternalPledgesControllerTest < ActionDispatch::IntegrationTest
 
   describe 'update method' do
     before do
-      @pledge = create :external_pledge, patient: @patient
+
+      @patient.external_pledges.create source: 'Baltimore Abortion Fund',
+                                     amount: 100,
+                                     created_by: @user
+      @pledge = @patient.external_pledges.first
+      #@pledge = create :external_pledge, patient: @patient
```
Comment (colinxfleming) [inline test/controllers/external_pledges_controller_test.rb]: Dead line, right?

File: test/controllers/external_pledges_controller_test.rb [hunk 2]
```diff
@@ -74,7 +79,13 @@ class ExternalPledgesControllerTest < ActionDispatch::IntegrationTest
   end
 
   describe 'destroy' do
-    before { @pledge = create :external_pledge, patient: @patient }
+    before do
+      @patient.external_pledges.create source: 'Baltimore Abortion Fund',
+                                     amount: 100,
+                                     created_by: @user
+      @pledge = @patient.external_pledges.first
+      #@pledge = create :external_pledge, patient: @patient
```
Comment (colinxfleming) [inline test/controllers/external_pledges_controller_test.rb]: Dead line?

File: test/system/pledge_fulfillment_test.rb [hunk 3]
```diff
@@ -15,7 +15,6 @@ class PledgeFulfillmentTest < ApplicationSystemTestCase
     @nonpledged_pt = create :patient, clinic: @clinic,
                                       appointment_date: 2.weeks.from_now,
                                       fund_pledge: 500
-    @fulfillment = create :fulfillment, patient: @pledged_pt
   end
```
Comment (colinxfleming) [inline test/system/pledge_fulfillment_test.rb]: Guessing that these got removed because we're creating fulfillment with an object callback? 👍 for cleaning up my bad test code, thank you!

Comment (lomky) [reply test/system/pledge_fulfillment_test.rb]: That seems a likely explanation! I just noticed it was unneeded on the console and gave deleting it in the test a shot. Still functioned!

File: test/controllers/external_pledges_controller_test.rb (hunk 2 above)
Comment (lomky) [reply test/controllers/external_pledges_controller_test.rb]: yup; will delete

File: test/controllers/external_pledges_controller_test.rb (hunk 1 above)
Comment (lomky) [reply test/controllers/external_pledges_controller_test.rb]: yep, will delete.

**Comment by lomky** [comment]
I am all for a couple days of testing this before we smash the merge. Our test suite is pretty sweet, but with a refactor this intrusive caution is the better part of valor. Let me know of any hands on testing you'd like me to do & document that would help clear up concerns on your part.

**Comment by colinxfleming** [comment]
I mostly just want to confirm that merging this doesn't change the json stored in db/alter our object model; if that's the case we'll want to write a db migration or run a script to generate the keys, but that'll be pretty trivial as it's something we've done a few times before. I think in general you can treat this as approved.

**Comment by lomky** [comment]
hmm... It technically does change the object model, it just doesn't act upon those changes. It _allows_ `call`s, `external_pledge`s, and `fulfillment`s to below to different parent objects, but then doesn't assign them to anything but `patient`s. We also decided that as far as the Controllers are concerned, it only has `patient` parent objects. I'll dig a bit on the docs and see if I can get a clear statement on how this will affect the db.

**Comment by colinxfleming** [comment]
hello from scenic nashville!

I had an hour to kill so I pulled this down and gave it a shot with a fresh db seed off of master. On both of these, I confirmed in the rails console that `calls`, `external_pledges`, and `fulfillments` maintain their current behavior.

* I modified Patient 8 by throwing on an ext pledge, call, and fulfillment. Here's what the object looks like in db, logging into the mongo shell:

```
> db.patients.find ( { "name": "Patient 8" } )
{ "_id" : ObjectId("5aabb910ac87fda0b83e47c4"), "special_circumstances" : [ ], "name" : "Patient 8", "primary_phone" : "1231231238", "initial_call_date" : ISODate("2018-03-13T00:00:00Z"), "urgent_flag" : false, "last_menstrual_period_weeks" : 10, "last_menstrual_period_days" : 3, "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "voicemail_preference" : "not_specified", "line" : "DC", "identifier" : "D3-1238", "updated_at" : ISODate("2018-03-16T12:38:52.651Z"), "created_at" : ISODate("2018-03-16T12:31:12.799Z"), "version" : 7, "fulfillment" : { "_id" : ObjectId("5aabb910ac87fda0b83e47c6"), "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:12.806Z"), "created_at" : ISODate("2018-03-16T12:31:12.806Z"), "version" : 1, "check_number" : "", "fulfilled" : true, "gestation_at_procedure" : "5", "procedure_cost" : 1000 }, "calls" : [ { "_id" : ObjectId("5aabb911ac87fda0b83e486c"), "status" : "Left voicemail", "created_at" : ISODate("2018-03-13T12:31:13.351Z"), "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:13.352Z"), "version" : 1 }, { "_id" : ObjectId("5aabb911ac87fda0b83e486f"), "status" : "Left voicemail", "created_at" : ISODate("2018-03-13T12:31:13.358Z"), "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:13.358Z"), "version" : 1 }, { "_id" : ObjectId("5aabb911ac87fda0b83e4872"), "status" : "Left voicemail", "created_at" : ISODate("2018-03-13T12:31:13.364Z"), "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:13.365Z"), "version" : 1 }, { "_id" : ObjectId("5aabb911ac87fda0b83e4875"), "status" : "Left voicemail", "created_at" : ISODate("2018-03-13T12:31:13.370Z"), "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:13.371Z"), "version" : 1 }, { "_id" : ObjectId("5aabb911ac87fda0b83e4878"), "status" : "Left voicemail", "created_at" : ISODate("2018-03-13T12:31:13.376Z"), "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:13.377Z"), "version" : 1 }, { "_id" : ObjectId("5aabbaaeac87fda2aa426778"), "status" : "Reached patient", "created_by_id" : ObjectId("5aabb90fac87fda0b83e4796"), "updated_by_id" : ObjectId("5aabb90fac87fda0b83e4796"), "updated_at" : ISODate("2018-03-16T12:38:06.091Z"), "created_at" : ISODate("2018-03-16T12:38:06.091Z"), "version" : 1 } ], "notes" : [ { "_id" : ObjectId("5aabb911ac87fda0b83e488f"), "full_text" : "This is a note This is a note This is a note This is a note This is a note This is a note This is a note This is a note This is a note This is a note ", "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:13.448Z"), "created_at" : ISODate("2018-03-16T12:31:13.448Z"), "version" : 1 }, { "_id" : ObjectId("5aabb911ac87fda0b83e4891"), "full_text" : "Additional note Additional note Additional note Additional note Additional note Additional note Additional note Additional note Additional note Additional note ", "created_by_id" : ObjectId("5aabb90fac87fda0b83e4798"), "updated_at" : ISODate("2018-03-16T12:31:13.452Z"), "created_at" : ISODate("2018-03-16T12:31:13.452Z"), "version" : 1 } ], "clinic_id" : ObjectId("5aabb910ac87fda0b83e479e"), "last_edited_by_id" : ObjectId("5aabb90fac87fda0b83e4796"), "referred_to_clinic" : false, "resolved_without_fund" : false, "updated_by_id" : ObjectId("5aabb90fac87fda0b83e4796"), "appointment_date" : ISODate("2018-03-21T00:00:00Z"), "procedure_cost" : 1000, "fund_pledge" : 100, "external_pledges" : [ { "_id" : ObjectId("5aabbacaac87fda2aa42677f"), "active" : true, "source" : "Clinic discount", "amount" : 30, "created_by_id" : ObjectId("5aabb90fac87fda0b83e4796"), "updated_by_id" : ObjectId("5aabb90fac87fda0b83e4796"), "updated_at" : ISODate("2018-03-16T12:38:34.761Z"), "created_at" : ISODate("2018-03-16T12:38:34.761Z"), "version" : 1 } ], "pledge_sent" : true, "pledge_sent_at" : ISODate("2018-03-16T12:38:43.346Z"), "pledge_sent_by_id" : ObjectId("5aabb90fac87fda0b83e4796") }
```

There's a ton of info in there, but it still has the appropriate keys in the json.

Similarly, for a new patient, here's the big old blob:

```
> db.patients.find ( { "name": "New polymorphic patient" } )
{
   "_id":ObjectId("5aabbb3aac87fda2aa426784"),
   "special_circumstances":[

   ],
   "name":"New polymorphic patient",
   "last_menstrual_period_days":null,
   "last_menstrual_period_weeks":null,
   "primary_phone":"5555555555",
   "line":"DC",
   "language":"",
   "voicemail_preference":"not_specified",
   "initial_call_date":   ISODate("2018-03-16T00:00:00   Z"),
   "created_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
   "identifier":"D5-5555",
   "updated_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
   "updated_at":   ISODate("2018-03-16T12:41:21.030   Z"),
   "created_at":   ISODate("2018-03-16T12:40:26.917   Z"),
   "version":7,
   "fulfillment":{
      "_id":ObjectId("5aabbb3aac87fda2aa426786"),
      "created_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
      "updated_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
      "updated_at":      ISODate("2018-03-16T12:40:26.928      Z"),
      "created_at":      ISODate("2018-03-16T12:40:26.928      Z"),
      "version":1,
      "check_number":"",
      "fulfilled":true,
      "gestation_at_procedure":"7",
      "procedure_cost":200
   },
   "user_ids":[
      ObjectId("5aabb90fac87fda0b83e4796")
   ],
   "appointment_date":   ISODate("2018-03-27T00:00:00   Z"),
   "last_edited_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
   "clinic_id":ObjectId("5aabb910ac87fda0b83e479c"),
   "referred_to_clinic":false,
   "resolved_without_fund":false,
   "procedure_cost":300,
   "fund_pledge":200,
   "external_pledges":[
      {
         "_id":ObjectId("5aabbb5bac87fda2aa42678d"),
         "active":true,
         "source":"Metallica Abortion Fund",
         "amount":30,
         "created_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
         "updated_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
         "updated_at":         ISODate("2018-03-16T12:40:59.809         Z"),
         "created_at":         ISODate("2018-03-16T12:40:59.809         Z"),
         "version":1
      }
   ],
   "calls":[
      {
         "_id":ObjectId("5aabbb62ac87fda2aa42678f"),
         "status":"Reached patient",
         "created_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
         "updated_by_id":ObjectId("5aabb90fac87fda0b83e4796"),
         "updated_at":         ISODate("2018-03-16T12:41:06.167         Z"),
         "created_at":         ISODate("2018-03-16T12:41:06.167         Z"),
         "version":1
      }
   ],
   "pledge_sent":true,
   "pledge_sent_at":   ISODate("2018-03-16T12:41:14.261   Z"),
   "pledge_sent_by_id":ObjectId("5aabb90fac87fda0b83e4796")
}
```

In short, I'm pretty satisfied that this is behaving properly. @lomky if you feel good about where this is I think we should merge it, play with it in sandbox for a little while, and then smash merge. thoughts?

**Comment by colinxfleming** [comment]
Having received clearance from @lomky on the game plan, I'm smashin that merge button!

Thank you @lomky for the excellent work implementing a great idea.
